File: 100/python3/main.py
#!/usr/bin/env python3

class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        p_stk, q_stk = [p], [q]

        while p_stk and q_stk:
            node_p = p_stk.pop()
            node_q = q_stk.pop()

            if not node_p and not node_q:
                continue

            if (not node_p or not node_q) or (node_p.val != node_q.val):
                return False

            p_stk.append(node_p.right)
            q_stk.append(node_q.right)
            p_stk.append(node_p.left)
            q_stk.append(node_q.left)

        return not p_stk and not q_stk

File: 100/cpp/main.cpp
#include <iostream>
#include <stack>

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        std::stack<TreeNode*> p_stk, q_stk;

        p_stk.push(p);
        q_stk.push(q);

        while (!p_stk.empty() && !q_stk.empty()) {
            TreeNode* node_p = p_stk.top();
            TreeNode* node_q = q_stk.top();
            p_stk.pop();
            q_stk.pop();

            if (node_p == nullptr && node_q == nullptr) continue;
            if (node_p == nullptr || node_q == nullptr) return false;
            if (node_p->val != node_q->val) return false;

            // Push right and left children of both nodes onto the stacks
            p_stk.push(node_p->right);
            q_stk.push(node_q->right);
            p_stk.push(node_p->left);
            q_stk.push(node_q->left);
        }

        return p_stk.empty() && q_stk.empty();
    }
};
